From 1afd7e5a036225f598e4852c47ed122ebfb25648 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Mon, 8 Apr 2024 16:56:09 +0800
Subject: [PATCH 103/265] mtk: mt76: mt7996: Fix get_txpower wrong result in
 single wiphy and legacy mode

Fix get_txpower wrong result in single wiphy and legacy mode.
ieee80211_hw is get from wiphy0, so we need to get correct phy from vif.

Temporarily use link 0 bss due to mac80211 didn't pass link id here.

---
 mt7996/main.c | 28 +++++++++++++++++++++++++++-
 1 file changed, 27 insertions(+), 1 deletion(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 9533de26..408f848e 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -981,6 +981,32 @@ out:
 	mutex_unlock(&dev->mt76.mutex);
 }
 
+int mt7996_get_txpower(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
+		       int *dbm)
+{
+	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
+	struct mt7996_bss_conf *mconf;
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
+	struct mt76_phy *mphy;
+	int delta;
+
+	mutex_lock(&dev->mt76.mutex);
+	mconf = mconf_dereference_protected(mvif, mvif->master_link_id);
+	if (!mconf || !mconf->phy) {
+		*dbm = 0;
+		goto out;
+	}
+
+	mphy = mconf->phy->mt76;
+
+	delta = mt76_tx_power_nss_delta(hweight16(mphy->chainmask));
+
+	*dbm = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
+out:
+	mutex_unlock(&dev->mt76.mutex);
+	return 0;
+}
+
 static void
 mt7996_channel_switch_beacon(struct ieee80211_hw *hw,
 			     struct ieee80211_vif *vif,
@@ -2672,7 +2698,7 @@ const struct ieee80211_ops mt7996_ops = {
 	.hw_scan = mt7996_hw_scan,
 	.cancel_hw_scan = mt7996_cancel_hw_scan,
 	.release_buffered_frames = mt76_release_buffered_frames,
-	.get_txpower = mt76_get_txpower,
+	.get_txpower = mt7996_get_txpower,
 	.channel_switch_beacon = mt7996_channel_switch_beacon,
 	.get_stats = mt7996_get_stats,
 	.get_et_sset_count = mt7996_get_et_sset_count,
-- 
2.45.2

